File: not.r/extractSeqFromFasta.pl
#!/usr/bin/perl
use strict;

if(@ARGV != 2){
	print("usage: fa gtf");
	die;
}

open IN,$ARGV[1] or die "cannot open $ARGV[1]: $!\n";
while(my $l=<IN>){
	next if(substr($l,0,1) eq '#');
	chomp $l;
	my @l = split('\t',$l);
	my @r = `samtools faidx $ARGV[0] $l[0]:$l[3]-$l[4]`;
	chomp @r;
	my $seq = join('',@r[1..$#r]);
	$seq = revCompl($seq) if($l[6] eq '-');
	print ">$l[8]\n".$seq."\n";
}

close IN;

sub revCompl{
	my $r = $_[0];
	$r =~ tr/ATGCatgc/TACGtacg/; 
	$r=reverse($r);
	$r
}

File: not.r/extractSeqFromFasta.py
#!/usr/bin/python
import os
import sys

#usage:cat *gtf | python  *.py  fasta [output format: empty for tab, 'fa' for fasta] > .fa
path2fa = sys.argv[1]
complement={'a':'t','t':'a','g':'c','c':'g','n':'n','A':'T','T':'A','G':'C','C':'G','N':'N'}

def getSeq(fa,seqname,start,stop,strand):
	start -= 1 #since gtf coordinates are 1-based
	if seqname in fa:
		r = fa[seqname][start:stop]
	else:
		return('-')
	if strand == '-':
		r = ''.join([complement[i] for i in r[::-1]])
	return(r)

def readFasta(f):
	fa = open(f,buffering=1000000)
	r = {}
	name = ''
	lines = []
	i = 0
	for line in fa:
		i += 1
		line = line.rstrip()
		if line[0:1] == '>':
			if name != '':
				r[name] = ''.join(lines)
			name = line[1:]
			lines = []
		else:
			lines.append(line)
	r[name] = ''.join(lines)
	fa.close()
	return(r)

fa = readFasta(path2fa)

for l in sys.stdin:
	l = l.rstrip()
	line = l.split("\t")
	seq =  getSeq(fa,line[0],int(line[3]),int(line[4]),line[6])
	if (len(sys.argv) == 3) & (sys.argv[2] == 'fa'):
		print(">"+line[8]+"\n"+seq)
	else:
		print(line[8]+"\t"+seq)
